File: framework/src/play-java-jdbc/src/main/java/play/db/DBModule.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.db;

import com.google.common.collect.ImmutableList;
import com.typesafe.config.Config;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.Environment;
import play.inject.Binding;
import play.inject.Module;

import javax.inject.Inject;
import javax.inject.Provider;
import java.util.List;
import java.util.Set;

/**
 * Injection module with default DB components.
 */
public final class DBModule extends Module {

    private static final Logger logger = LoggerFactory.getLogger(DBModule.class);

    @Override
    public List<Binding<?>> bindings(final Environment environment, final Config config) {
        String dbKey = config.getString("play.db.config");
        String defaultDb = config.getString("play.db.default");

        ImmutableList.Builder<Binding<?>> list = new ImmutableList.Builder<Binding<?>>();

        list.add(bindClass(ConnectionPool.class).to(DefaultConnectionPool.class));
        list.add(bindClass(DBApi.class).to(DefaultDBApi.class));

        try {
            Set<String> dbs = config.getConfig(dbKey).root().keySet();
            for (String db : dbs) {
                list.add(bindClass(Database.class).qualifiedWith(named(db)).to(new NamedDatabaseProvider(db)));
            }

            if (dbs.contains(defaultDb)) {
                list.add(bindClass(Database.class).to(bindClass(Database.class).qualifiedWith(named(defaultDb))));
            }
        } catch (com.typesafe.config.ConfigException.Missing ex) {
            logger.warn("Configuration not found for database: {}", ex.getMessage());
        }

        return list.build();
    }

    private NamedDatabase named(String name) {
        return new NamedDatabaseImpl(name);
    }

    /**
     * Inject provider for named databases.
     */
    public static class NamedDatabaseProvider implements Provider<Database> {
        @Inject private DBApi dbApi = null;
        private final String name;

        public NamedDatabaseProvider(String name) {
            this.name = name;
        }

        public Database get() {
            return dbApi.getDatabase(name);
        }
    }

}

File: framework/src/play-jdbc/src/main/scala/play/api/db/DBModule.scala
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.db

import com.typesafe.config.Config
import javax.inject.{ Inject, Provider, Singleton }
import play.api._
import play.api.inject._
import play.db.NamedDatabaseImpl

import scala.concurrent.Future
import scala.util.Try

/**
 * DB runtime inject module.
 */
final class DBModule extends SimpleModule((environment, configuration) => {
  def bindNamed(name: String): BindingKey[Database] = {
    bind[Database].qualifiedWith(new NamedDatabaseImpl(name))
  }

  def namedDatabaseBindings(dbs: Set[String]): Seq[Binding[_]] = dbs.toSeq.map { db =>
    bindNamed(db).to(new NamedDatabaseProvider(db))
  }

  def defaultDatabaseBinding(default: String, dbs: Set[String]): Seq[Binding[_]] = {
    if (dbs.contains(default)) Seq(bind[Database].to(bindNamed(default))) else Nil
  }

  val dbKey = configuration.underlying.getString("play.db.config")
  val default = configuration.underlying.getString("play.db.default")
  val dbs = configuration.getOptional[Configuration](dbKey).getOrElse(Configuration.empty).subKeys
  Seq(
    bind[DBApi].toProvider[DBApiProvider]
  ) ++ namedDatabaseBindings(dbs) ++ defaultDatabaseBinding(default, dbs)
})

/**
 * DB components (for compile-time injection).
 */
trait DBComponents {
  def environment: Environment
  def configuration: Configuration
  def connectionPool: ConnectionPool
  def applicationLifecycle: ApplicationLifecycle

  lazy val dbApi: DBApi = new DBApiProvider(environment, configuration, connectionPool, applicationLifecycle, None).get
}

/**
 * Inject provider for DB implementation of DB API.
 */
@Singleton
class DBApiProvider(
    environment: Environment,
    configuration: Configuration,
    defaultConnectionPool: ConnectionPool,
    lifecycle: ApplicationLifecycle,
    maybeInjector: Option[Injector]
) extends Provider[DBApi] {

  @Inject
  def this(
    environment: Environment,
    configuration: Configuration,
    defaultConnectionPool: ConnectionPool,
    lifecycle: ApplicationLifecycle,
    injector: Injector = NewInstanceInjector
  ) = {
    this(environment, configuration, defaultConnectionPool, lifecycle, Option(injector))
  }

  lazy val get: DBApi = {
    val config = configuration.underlying
    val dbKey = config.getString("play.db.config")
    val pool = maybeInjector
      .map(injector => ConnectionPool.fromConfig(config.getString("play.db.pool"), injector, environment, defaultConnectionPool))
      .getOrElse(ConnectionPool.fromConfig(config.getString("play.db.pool"), environment, defaultConnectionPool))
    val configs = if (config.hasPath(dbKey)) {
      Configuration(config).getPrototypedMap(dbKey, "play.db.prototype").mapValues(_.underlying)
    } else Map.empty[String, Config]
    val db = new DefaultDBApi(configs, pool, environment, maybeInjector.getOrElse(NewInstanceInjector))
    lifecycle.addStopHook { () => Future.fromTry(Try(db.shutdown())) }
    db.initialize(logInitialization = environment.mode != Mode.Test)
    db
  }
}

/**
 * Inject provider for named databases.
 */
class NamedDatabaseProvider(name: String) extends Provider[Database] {
  @Inject private var dbApi: DBApi = _
  lazy val get: Database = dbApi.database(name)
}
